File: src/main/java/org/stt/command/CommandFormatter.java
package org.stt.command;

import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.TokenStream;
import org.stt.grammar.EnglishCommandsLexer;
import org.stt.grammar.EnglishCommandsParser;
import org.stt.model.TimeTrackingItem;
import org.stt.time.DateTimes;

import javax.inject.Inject;
import javax.inject.Named;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Objects;

public class CommandFormatter {
    private final CommandTextParser commandTextParser;
    private final DateTimeFormatter dateTimeFormatter;
    private final DateTimeFormatter timeFormatter;

    @Inject
    public CommandFormatter(CommandTextParser commandTextParser,
                            @Named("dateTimeFormatter") DateTimeFormatter dateTimeFormatter,
                            @Named("timeFormatter") DateTimeFormatter timeFormatter) {
        this.commandTextParser = commandTextParser;
        this.dateTimeFormatter = dateTimeFormatter;
        this.timeFormatter = timeFormatter;
    }

    public Command parse(String command) {
        Objects.requireNonNull(command);
        CharStream inputStream = new CaseInsensitiveInputStream(command);
        EnglishCommandsLexer lexer = new EnglishCommandsLexer(inputStream);
        TokenStream tokenStream = new CommonTokenStream(lexer);
        EnglishCommandsParser parser = new EnglishCommandsParser(tokenStream);
        Object result = commandTextParser.walk(parser.command());
        if (result instanceof TimeTrackingItem) {
            TimeTrackingItem parsedItem = (TimeTrackingItem) result;
            return new NewActivity(parsedItem);
        }
        if (result instanceof LocalDateTime) {
            return new EndCurrentItem((LocalDateTime) result);
        }
        if (result instanceof Command) {
            return (Command) result;
        }
        return DoNothing.INSTANCE;
    }

    public String asNewItemCommandText(TimeTrackingItem item) {
        Objects.requireNonNull(item);
        String start;
        if (DateTimes.isToday(item.getStart())) {
            start = timeFormatter.format(item.getStart().toLocalTime());
        } else {
            start = dateTimeFormatter.format(item.getStart());
        }
        return item.getEnd()
                .map(endDateTime -> {
                            String end;
                            if (DateTimes.isToday(endDateTime)) {
                                end = timeFormatter.format(endDateTime.toLocalTime());
                            } else {
                                end = dateTimeFormatter.format(endDateTime);
                            }
                            return String.format("%s from %s to %s", item.getActivity(), start, end);
                        }
                )
                .orElseGet(() -> String.format("%s since %s", item.getActivity(), start));
    }

}

File: src/main/kotlin/org/stt/command/CommandFormatter.kt
package org.stt.command

import org.antlr.v4.runtime.CommonTokenStream
import org.stt.grammar.EnglishCommandsLexer
import org.stt.grammar.EnglishCommandsParser
import org.stt.model.TimeTrackingItem
import org.stt.time.DateTimes
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import javax.inject.Inject
import javax.inject.Named

class CommandFormatter @Inject
constructor(private val commandTextParser: CommandTextParser,
            @param:Named("dateTimeFormatter") private val dateTimeFormatter: DateTimeFormatter,
            @param:Named("timeFormatter") private val timeFormatter: DateTimeFormatter) {

    fun parse(command: String): Command {
        val inputStream = CaseInsensitiveInputStream(command)
        val lexer = EnglishCommandsLexer(inputStream)
        val tokenStream = CommonTokenStream(lexer)
        val parser = EnglishCommandsParser(tokenStream)
        val result = commandTextParser.walk(parser.command())
        if (result is TimeTrackingItem) {
            return NewActivity(result)
        }
        return if (result is LocalDateTime) {
            EndCurrentItem(result)
        } else result as? Command ?: DoNothing
    }

    fun asNewItemCommandText(item: TimeTrackingItem): String {
        val start = if (DateTimes.isToday(item.start)) {
            timeFormatter.format(item.start.toLocalTime())
        } else {
            dateTimeFormatter.format(item.start)
        }
        return item.end?.let { endDateTime ->
            val end = if (DateTimes.isToday(endDateTime)) {
                timeFormatter.format(endDateTime.toLocalTime())
            } else {
                dateTimeFormatter.format(endDateTime)
            }
            String.format("%s from %s to %s", item.activity, start, end)
        } ?: String.format("%s since %s", item.activity, start)
    }

}
